### PR TITLE
83 assignment I and II job service

### PR DESCRIPTION
**Pull request by SanFenDevelop**
### Major update includes Assignment 1 and 2, mainly changes in the port and core packages. 

completed following tasks:

**Assignment 1:**

1.  Create a folder for your service in the services folder
2.  Copy folder structure from entity service into your service
3.  Rename/Remove entity specifics from your service

**Assignment 2:**

1. Implement api.go and other ports
2. Implement model.go
3. Implement core including unit tests

_Improvements are certainly still possible, a merge simplifies the processing of Assignment 3 and 4_

### CODE REVIEW

File: services/job/go.mod [hunk 1]
```diff
@@ -0,0 +1,8 @@
+module github.com/informatik-mannheim/cmg-ss2025/services/job
+
+go 1.23.1
+
+require (
+	github.com/google/uuid v1.6.0
```
Comment (Tonka-de) [inline services/job/go.mod:6]: is it approved though?

File: services/job/main.go [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+		log.Print("The service is shutting down...")
+		srv.Shutdown(context.Background())
+	}()
+
+	log.Print("listening...")
+	srv.ListenAndServe()
+	log.Print("Done")
+}
+*/
```
Comment (Tonka-de) [inline services/job/main.go:26]: Why comment out the main?

File: services/job/adapters/handler-http/handler.go [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusNotFound)
+		return
+	}
+	w.Header().Set("Content-Type", "application/json")
+	json.NewEncoder(w).Encode(job)
+}
+
+
+*/
```
Comment (Tonka-de) [inline services/job/adapters/handler-http/handler.go:51]: Why is this commented out?

File: services/job/adapters/repo-in-memory/repo.go [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+func (r *Repo) FindById(id string, ctx context.Context) (ports.Job, error) {
+	job, ok := r.entities[id]
+	if !ok {
+		return ports.Job{}, ports.ErrJobNotFound
+	}
+	return job, nil
+}
+
+*/
```
Comment (Tonka-de) [inline services/job/adapters/repo-in-memory/repo.go:30]: Why is this commented out?

File: services/job/core/job_test.go [hunk 5]
```diff
@@ -0,0 +1,599 @@
...
+	if !ok {
+		return ports.Job{}, errors.New("job not found")
+	}
+
+	if err := updateFn(&job); err != nil {
+		return ports.Job{}, err
+	}
+	m.jobs[id] = job
+	return job, nil
+}
```
Comment (Tonka-de) [inline services/job/core/job_test.go:59]: This is technically an adapter which in turn should be in its own file in /adapters

File: services/job/core/job_test.go [hunk 6]
```diff
@@ -0,0 +1,599 @@
...
+
+func containsStatus(statusList []ports.JobStatus, status ports.JobStatus) bool {
```
Comment (Tonka-de) [inline services/job/core/job_test.go:61]: This sounds like a utility function, maybe wrong place to be in?

File: services/job/core/job_test.go [hunk 7]
```diff
@@ -0,0 +1,599 @@
...
+	for _, s := range statusList {
+		if s == status {
+			return true
+		}
+	}
```
Comment (Tonka-de) [inline services/job/core/job_test.go:66]: VSCode tells me this can be simplyfied with slices.Contains

File: services/job/core/job_test.go [hunk 8]
```diff
@@ -0,0 +1,599 @@
...
+			id:   createdJob.Id,
+			data: ports.SchedulerUpdateData{
+				WorkerID:        uuid.NewString(),
+				ComputeZone:     "FR",
+				CarbonIntensity: -1, // Negative value
+				CarbonSaving:    30,
+				Status:          ports.StatusScheduled,
+			},
+			wantErr: true,
+		},
```
Comment (Tonka-de) [inline services/job/core/job_test.go:482]: Do you also test for negative CarbonSaving? Dont know if that should be allowed or not, just bringing it up

File: services/job/core/job_test.go [hunk 9]
```diff
@@ -0,0 +1,599 @@
...
+		{
+			name: "Scheduled status without Result",
+			id:   createdJob.Id,
+			data: ports.WorkerDaemonUpdateData{
+				Status:       ports.StatusScheduled,
+				Result:       "",
+				ErrorMessage: "",
+			},
+			wantErr: false, // Assuming this is acceptable; adjust according to your logic
```
Comment (Tonka-de) [inline services/job/core/job_test.go:579]: scheduled is actually meant to be set from the Job-Scheduler, not the worker daemon

File: services/job/core/job_test.go [hunk 10]
```diff
@@ -0,0 +1,599 @@
...
+			}
+		})
+	}
+}
+
+func generateLargeParameters(n int) map[string]string {
+	params := make(map[string]string)
+	for i := 0; i < n; i++ {
+		params[fmt.Sprintf("param%d", i)] = fmt.Sprintf("value%d", i)
+	}
```
Comment (Tonka-de) [inline services/job/core/job_test.go:597]: vscode tells me this can be modernized with "range", consider using it

File: services/job/core/job.go [hunk 11]
```diff
@@ -0,0 +1,227 @@
...
+)
+
+// JobService is a struct that implements the JobService interface.
+// It provides methods to manage jobs, including creating, retrieving, and updating jobs.
+// It uses a storage interface to interact with the underlying data store.
+type JobService struct {
+	storage ports.JobStorage
+}
+
+// NewJobService creates a new instance of JobService with the provided storage.
```
Comment (Tonka-de) [inline services/job/core/job.go:21]: In my opinion (I will write this only once), there are some comments that do nothing, not even help and make everything look more complicated. This is a personal opinion, but I would remove comments like these. Sometimes less = more

File: services/job/core/job.go [hunk 12]
```diff
@@ -0,0 +1,227 @@
...
+func NewJobService(storage ports.JobStorage) *JobService {
+	return &JobService{
+		storage: storage,
+	}
+}
+
```
Comment (Tonka-de) [inline services/job/core/job.go:27]: should add an check if its actually implemented, if memory serves right that is possible via `var _ ports.JobService = JobService{nil}` or something like that

File: services/job/core/job.go [hunk 13]
```diff
@@ -0,0 +1,227 @@
...
+// GetJobs retrieves jobs based on their status.
+// It returns a slice of jobs that match the provided status.
+// If no status is provided, it returns all jobs.
+func (s *JobService) GetJobs(ctx context.Context, status []ports.JobStatus) ([]ports.Job, error) {
+	// If status is nil, return all jobs
+	if status == nil {
```
Comment (Tonka-de) [inline services/job/core/job.go:33]: for now ig this is fine, but should be reconsidered at least when adding auth

File: services/job/core/job.go [hunk 14]
```diff
@@ -0,0 +1,227 @@
...
+	// Input validation
+	if strings.TrimSpace(jobCreate.JobName) == "" {
+		return ports.Job{}, errors.New("job name must be provided")
+	}
+	if strings.TrimSpace(jobCreate.CreationZone) == "" {
+		return ports.Job{}, errors.New("creation zone must be provided")
+	}
+	if strings.TrimSpace(jobCreate.Image.Name) == "" || strings.TrimSpace(jobCreate.Image.Version) == "" {
+		return ports.Job{}, errors.New("both image name and version must be provided")
+	}
```
Comment (Tonka-de) [inline services/job/core/job.go:66]: personally, I would make validation functions for this to be reused thoughout the service, going by DRY (dont repeat yourself). Should definetly changed when adding auth, for now its fine ig

File: services/job/core/job.go [hunk 15]
```diff
@@ -0,0 +1,227 @@
...
+
+	// Version format validation without regex
+	if !isSimpleValidVersion(jobCreate.Image.Version) {
+		return ports.Job{}, errors.New("image version format is invalid")
+	}
+
+	if len(jobCreate.Parameters) == 0 {
```
Comment (Tonka-de) [inline services/job/core/job.go:73]: must there though? When having 0 parameters I think you would get something like `docker run ImageName`, which would be just fine, valid command

**Review by Tonka-de — CHANGES_REQUESTED** [review]
Overall: not bad, ~90% test coverage, tests work but I could not run the service if I tried to start it with `go run`. The use of an unapproved package is also not good. Then, (if the package has it that is), you should consider switching all those "id-strings" to UUID types, saves you the validation ig (except if that is by design).

Last question: why is there a **Makefile**?

**Comment by Tonka-de** [comment]
1 more thing I forgot: please merge main into your branch

File: services/job/main.go (hunk 2 above)
Comment (SanFenDevelop) [reply services/job/main.go:26]: Since the service itself does not yet work, the aim was to create the stubs and the business logic with tests. The main function will be implemented in Assessment 3, i.e. in the next PR

File: services/job/adapters/handler-http/handler.go (hunk 3 above)
Comment (SanFenDevelop) [reply services/job/adapters/handler-http/handler.go:51]: The same reason as for the Main function

File: services/job/core/job_test.go (hunk 5 above)
Comment (SanFenDevelop) [reply services/job/core/job_test.go:59]: You are right in principle, but as the adapter was only created for the test, I was unsure and therefore left it in the same file

File: services/job/ports/repo.go [hunk 16]
```diff
@@ -0,0 +1,12 @@
...
+
+import (
+	"context"
+)
+
+type JobStorage interface {
+	GetJobs(ctx context.Context, status []JobStatus) ([]Job, error)
+	CreateJob(ctx context.Context, job Job) error
+	GetJob(ctx context.Context, id string) (Job, error)
+	UpdateJob(ctx context.Context, id string, updateFn func(*Job) error) (Job, error)
```
[the remark this answers is not in the document]
Comment (SanFenDevelop) [reply services/job/ports/repo.go:11]: The idea is that the update function is intended for PATCH requests and is assigned an appropriate function depending on whether it is requested by the scheduler or worker daemon

**Comment by SanFenDevelop** [comment]
> Overall: not bad, ~90% test coverage, tests work but I could not run the service if I tried to start it with `go run`. The use of an unapproved package is also not good. Then, (if the package has it that is), you should consider switching all those "id-strings" to UUID types, saves you the validation ig (except if that is by design).
> 
> Last question: why is there a **Makefile**?

I have explained most of the concerns in the comments, the Makefile is just a stub for a later phase and has no function yet

-> Ill do the Merge as mentioned, thanks for your effort by reviewing this

File: services/job/go.mod (hunk 1 above)
Comment (Tonka-de) [reply services/job/go.mod:6]: Appearently it is, but its in a newer Main than you got, so update it ^^

File: services/job/core/job_test.go (hunk 5 above)
Comment (Tonka-de) [reply services/job/core/job_test.go:59]: I did talk with the prof about it, as far as I understood it really is an adapter, so it should probably also go there

File: services/job/ports/repo.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (Tonka-de) [reply services/job/ports/repo.go:11]: But shouldnt that also just be possible by using only `update Job` instead of `updateFn func(*Job) error`?

File: services/job/core/job_test.go (hunk 7 above)
Comment (SanFenDevelop) [reply services/job/core/job_test.go:66]: your idea requires an external package. I would like to avoid this if possible, thanks for this suggestion

File: services/job/core/job_test.go (hunk 6 above)
Comment (SanFenDevelop) [reply services/job/core/job_test.go:61]: created new util folder for specific function which could use multiple times

File: services/job/core/job_test.go (hunk 5 above)
Comment (SanFenDevelop) [reply services/job/core/job_test.go:59]: Iihave moved it to an adapter

File: services/job/go.mod (hunk 1 above)
Comment (SanFenDevelop) [reply services/job/go.mod:6]: alright thank you

File: services/job/adapters/repo-in-memory/mock_repo.go [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+	"errors"
+
+	"github.com/informatik-mannheim/cmg-ss2025/services/job/ports"
+	"github.com/informatik-mannheim/cmg-ss2025/services/job/utils"
+)
+
+// MockJobStorage is a mock implementation of the JobStorage interface
+type MockJobStorage struct {
+	jobs map[string]ports.Job
+}
```
Comment (Tonka-de) [inline services/job/adapters/repo-in-memory/mock_repo.go:14]: No check if it actually implements the interface

File: services/job/core/job_test.go (hunk 8 above)
Comment (SanFenDevelop) [reply services/job/core/job_test.go:426]: negative carbon savings could theoretically occur if no cheaper provider is available at the time and execution would also not be possible at the location of the order. It would then be possible for the compute zone to have a higher output than the creation zone, leading to negative carbon savings.

In my view, negative values should therefore be permitted

File: services/job/utils/job_utils.go [hunk 18]
```diff
@@ -0,0 +1,16 @@
+package utils
+
+import (
+	"github.com/informatik-mannheim/cmg-ss2025/services/job/ports"
+)
+
+// containsStatus checks if a given status is present in the status list.
+// It returns true if the status is found, otherwise false.
+func ContainsStatus(statusList []ports.JobStatus, status ports.JobStatus) bool {
```
Comment (Tonka-de) [inline services/job/utils/job_utils.go:9]: Isnt the status-list like a constant? Does it make sense to have it as parameter?

**Review by Tonka-de — CHANGES_REQUESTED** [review]
Currently in train, cant check over the testfile, please excuse this

File: services/job/core/job_test.go (hunk 9 above)
Comment (SanFenDevelop) [reply services/job/core/job_test.go:523]: yes that really makes little sense, i have replaced it with a test in which the error message must not be empty in the event of a failed job

File: services/job/core/job.go (hunk 12 above)
Comment (SanFenDevelop) [reply services/job/core/job.go:27]: Iihave adopted a good idea

File: services/job/core/job.go (hunk 14 above)
Comment (SanFenDevelop) [reply services/job/core/job.go:66]: you're right, it's currently only for the raw implementation, the service should be running first. I've made a note of that

File: services/job/core/job_test.go (hunk 7 above)
Comment (Tonka-de) [reply services/job/core/job_test.go]: That is not an external package, its in the std-lib

File: services/job/core/job_test.go (hunk 8 above)
Comment (Tonka-de) [reply services/job/core/job_test.go:426]: I dont think so, since a requirement is, that it must be positive. So -1 should be the only negative number indicating it is not set, otherwise it should always be positive

File: services/job/ports/repo.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (SanFenDevelop) [reply services/job/ports/repo.go:11]: yes, you're right, that makes the function unnecessarily complicated. I have replaced it with a job that simply sets the fields that are necessary.

File: services/job/core/job.go [hunk 19]
```diff
@@ -0,0 +1,227 @@
...
+func (s *JobService) GetJob(ctx context.Context, id string) (ports.Job, error) {
+	// Check for empty or whitespace-only ID
+	if len(strings.TrimSpace(id)) == 0 {
+		return ports.Job{}, errors.New("job ID must be provided")
+	}
+
+	// Validate UUID format
+	if _, err := uuid.Parse(id); err != nil {
+		return ports.Job{}, errors.New("job ID must be a valid UUID")
+	}
```
[the remark this answers is not in the document]
Comment (SanFenDevelop) [reply services/job/core/job.go:113]: I've made a note of it, will fix it as soon as I have time

File: services/job/core/job.go (hunk 15 above)
Comment (SanFenDevelop) [reply services/job/core/job.go:73]: right I just tried it, it automatically takes the :latest version from the Dockerhub

-removed that case, empty version should be fine

Comment (SanFenDevelop) [reply services/job/core/job.go:73]: that's right i should offer this option, will be deleted

**Comment by SanFenDevelop** [comment]
Thanks for your many comments and suggestions, I have applied a lot (most) of them and simplified things where possible. I think that a lot of edge cases have not yet been considered and the code needs further work. I think for now I should start with the adapters so that my application is already running and accessible. The tests are all green according to the latest status. Cheers
